### ISSUE TITLE
Investigate NaNs in FlashAttention

### ISSUE THREAD

**Issue by drisspg** [pytorch/pytorch#121558]
# Summary

NaNs for large bfloat16 inputs:

``` Python
import torch
from torch.nn.attention import SDPBackend, sdpa_kernel


query = torch.full((1,1, 1, 64), 133120.0, dtype=torch.bfloat16, device="cuda")
key = torch.full((1,1, 1, 64), 133120.0, dtype=torch.bfloat16, device="cuda")
value = torch.full((1,1, 1, 64), 133120.0, dtype=torch.bfloat16, device="cuda")

with sdpa_kernel(SDPBackend.FLASH_ATTENTION):
    out = torch.nn.functional.scaled_dot_product_attention(query, key, value)

print(out)
```
This produces NaNs for FA but the correct result for both efficient attention and the math path

cc @ezyang @gchanan @zou3519 @kadeng @jbschlosser @bhosmer @cpuhrsch @erichan1 @mikaylagawarecki

**Comment by LamForest** [comment]
I can reproduce this on latest commit:

testing code:
```python
import torch
from torch.nn.attention import SDPBackend, sdpa_kernel


query = torch.full((1,1, 1, 64), 133120.0, dtype=torch.bfloat16, device="cuda")
key = torch.full((1,1, 1, 64), 133120.0, dtype=torch.bfloat16, device="cuda")
value = torch.full((1,1, 1, 64), 133120.0, dtype=torch.bfloat16, device="cuda")

# my build disable CUDNN, so I skip CUDNN_ATTENTION
for backend in [SDPBackend.MATH, SDPBackend.FLASH_ATTENTION, SDPBackend.EFFICIENT_ATTENTION]:
    with sdpa_kernel(backend):
        print("backend = ", backend)
        out = torch.nn.functional.scaled_dot_product_attention(query, key, value)
        print(out)
```

terminal out:
```
➜  gtl_tmp python mha.py 
backend =  SDPBackend.MATH
tensor([[[[133120., 133120., 133120., 133120., 133120., 133120., 133120.,
           133120., 133120., 133120., 133120., 133120., 133120., 133120.,
           133120., 133120., 133120., 133120., 133120., 133120., 133120.,
           133120., 133120., 133120., 133120., 133120., 133120., 133120.,
           133120., 133120., 133120., 133120., 133120., 133120., 133120.,
           133120., 133120., 133120., 133120., 133120., 133120., 133120.,
           133120., 133120., 133120., 133120., 133120., 133120., 133120.,
           133120., 133120., 133120., 133120., 133120., 133120., 133120.,
           133120., 133120., 133120., 133120., 133120., 133120., 133120.,
           133120.]]]], device='cuda:0', dtype=torch.bfloat16)
backend =  SDPBackend.FLASH_ATTENTION
tensor([[[[nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
           nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
           nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan]]]],
       device='cuda:0', dtype=torch.bfloat16)
backend =  SDPBackend.EFFICIENT_ATTENTION
tensor([[[[133120., 133120., 133120., 133120., 133120., 133120., 133120.,
           133120., 133120., 133120., 133120., 133120., 133120., 133120.,
           133120., 133120., 133120., 133120., 133120., 133120., 133120.,
           133120., 133120., 133120., 133120., 133120., 133120., 133120.,
           133120., 133120., 133120., 133120., 133120., 133120., 133120.,
           133120., 133120., 133120., 133120., 133120., 133120., 133120.,
           133120., 133120., 133120., 133120., 133120., 133120., 133120.,
           133120., 133120., 133120., 133120., 133120., 133120., 133120.,
           133120., 133120., 133120., 133120., 133120., 133120., 133120.,
           133120.]]]], device='cuda:0', dtype=torch.bfloat16)
```

**Comment by drisspg** [comment]
Verified that with 2.2.1 Same output behavior, so I don't think this is related to the recent update to v2.5.5

**Comment by drisspg** [comment]
@tridao Thank you for looking into this! This example did come up from a user's model. I replaced the line you linked with 

`tensor(mi, ni) = exp2f(__fmul_rn(tensor(mi, ni), scale) - max_scaled);`

And I am observing the correct result. Do you think would be sufficient in all cases? Also if we want to avoid ffma should the change be holistic, e.g. `exp(tensor(mi, ni) - max_scaled)` and remove the log2(e) scaling? 

If so I do think it make sense to make this gated by a macro that PT can turn on and Flash can leave off

**Comment by tridao** [comment]
Yeah I think that should fix the problem. log2(e) scaling is not an issue, since it's equivalent to having a new_softmax_scale = softmax_scale * log2(e). The issue is I was trying to be clever and do the max before scaling, then combine (scale & subtract by max) into 1 operation (fma).

Long term I think it might be safer to do the softmax scaling on Q before multiplying with K.

For now we can have a macro to use __fmul_rn or not for this operation.

**Comment by Chillee** [comment]
To elaborate on @tridao's comment on why fma causes issues, in softmax (since I didn't quite follow why fma, an operator with *more precision* could cause problems).

We're computing `exp(x_i * scale - max_scaled)`. Now, `max_scaled = max(x_i * scale)`. The idea here is that before we take the exponent, we normalize everything down to 0 so that `exp(x)` doesn't blow up massively.

Now, for the largest value of `x_i`, `max_scaled = x_i * scale`. So, `x_i * scale - max_scaled == 0`, right?

Unfortunately, no, due to fma. `max_scaled` is actually equal to `round_fp32(x_i * scale)`. But in `fma`, `x_i * scale` never gets rounded! So with fma, we are computing `x_i * scale - round_fp32(x_i * scale)`, which is not equal to 0!

However, without fma, we would round `x_i * scale` to fp32 before subtracting it, and so we would get `round_fp32(x_i * scale) - round_fp32(x_i * scale)`, which *is* equal to 0.

**Comment by lezcano** [comment]
In this particular implementation, I'd consider benchmarking `fma` vs adding a `max(- , 0)` before the `exp`, which would effectively also make the computation numerically stable.

**Comment by Chillee** [comment]
@lezcano Bert (@bertmaher) mentioned that you can actually have an error in the other way too (i.e. a large negative error). So that would lead to a vector with all 0s, which since this is softmax, can lead to nans if you're then dividing by the sum.
